presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -518,6 +504,32 @@ public boolean transitionToFailed(Throwable throwable)
         return failed;
     }
 
+    public boolean transitionToCanceled()
+    {
+        recordDoneStats();
+
+        boolean canceled = queryState.setIf(FAILED, currentState -> !currentState.isDone());
+        if (canceled) {
+            failureCause.compareAndSet(null, toFailure(new PrestoException(USER_CANCELED, "Query was canceled")));
+            session.getTransactionId().ifPresent(autoCommit ? transactionManager::asyncAbort : transactionManager::fail);
```

Comment:
Is it a bug that `transitionToFailed` does this even if the `compareAndSet` fails?